### PR TITLE
Propagate show/hide window calls against the ConPTY pseudo window to the Terminal

### LINKED ISSUE #12570: Show/Hide Operations on ::GetConsoleWindow() do not work against Terminal through ConPTY
In attempting to make the Windows Terminal be the default terminal (#7414), we've discovered in testing that there are a lot of scenarios that were relying on manipulating the window handle from `::GetConsoleWindow()` in one way or another to show or hide the window state.

While we strongly discourage this as it isn't guaranteed to work in a universal way across platforms and systems... and is one of the APIs we have listed as "deprecated" on `docs.microsoft.com` for that exact reason... we're going to make a compromise solution here to try to make it work in enough scenarios that it will maintain compatibility with existing tooling that is relying on it. (One of the primary tenets of moving to Virtual Terminal sequences and the ConPTY is to make everything work in the "universal language" that all platforms support, so we strongly resist doing things "for Windows only".)

It looks like **XTWINOPS** (see https://invisible-island.net/xterm/ctlseqs/ctlseqs.html) is an existing sequence that will allow us to translate the window operations coming in and send them to the remote Terminal window. 

This may also work for some of the z-ordering issues of #2988 going forward.

The test case here specifically is doing:

```
::ShowWindow(::GetConsoleWindow(), SW_HIDE);
```

Or one of the other `SW_*` actions.

### PR DESCRIPTION
**Pull request by miniksa**
Propagate show/hide window calls against the ConPTY pseudo window to the Terminal

## PR Checklist
* [x] Closes #12570 
* [x] I work here
* [x] Manual Tests passed
* [x] Spec Link: →[Doc Link](https://github.com/microsoft/terminal/blob/dev/miniksa/msgs/doc/specs/%2312570%20-%20Show%20Hide%20operations%20on%20GetConsoleWindow%20via%20PTY.md)←

## Detailed Description of the Pull Request / Additional comments
- See the spec. It's pretty much everything I went through deciding on this.

## Validation Steps Performed
- [x] Manual validation against scratch application calling all of the `::ShowWindow` commands against the pseudo console "fake window" and observing the real terminal window state

### CODE REVIEW

File: src/interactivity/base/InteractivityFactory.cpp [hunk 1]
```diff
@@ -381,4 +390,79 @@ using namespace Microsoft::Console::Interactivity;
...
+    if (pFactory != nullptr)
+    {
+        return pFactory->PseudoWindowProc(hWnd, Message, wParam, lParam);
+    }
+
+    // If we get this far, call the default window proc
+    return DefWindowProcW(hWnd, Message, wParam, lParam);
+}
+
+[[nodiscard]] LRESULT CALLBACK InteractivityFactory::PseudoWindowProc(_In_ HWND hWnd, _In_ UINT Message, _In_ WPARAM wParam, _In_ LPARAM lParam)
```
Comment (miniksa) [inline src/interactivity/base/InteractivityFactory.cpp:421]: If you all want me to, I'll pull the `PseudoWindowProc` and window creation and whatnot into its own class file. But if you all don't care... then I'll leave it here.

File: src/cascadia/TerminalCore/Terminal.cpp [hunk 2]
```diff
@@ -221,6 +221,21 @@ void Terminal::UpdateAppearance(const ICoreAppearance& appearance)
...
+// //         we generate and send it here.
+// // Arguments:
+// // - visible: True for visible; false for not visible.
+// // Return Value:
+// // - <none>
+// void Terminal::UpdateVisibility(const bool visible)
+// {
+//     ReturnResponse(visible ? L"\x1b[1t" : L"\x1b[2t");
+// }
+
```
Comment (miniksa) [inline src/cascadia/TerminalCore/Terminal.cpp:238]: Trim commented code.

Comment (miniksa) [inline src/cascadia/TerminalCore/Terminal.cpp:238]: Trim commented code.

File: src/cascadia/TerminalCore/Terminal.hpp [hunk 3]
```diff
@@ -75,6 +75,7 @@ class Microsoft::Terminal::Core::Terminal final :
 
     void UpdateSettings(winrt::Microsoft::Terminal::Core::ICoreSettings settings);
     void UpdateAppearance(const winrt::Microsoft::Terminal::Core::ICoreAppearance& appearance);
+    // void UpdateVisibility(const bool visible);
```
Comment (miniksa) [inline src/cascadia/TerminalCore/Terminal.hpp:78]: Trim commented code.

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 4]
```diff
@@ -538,13 +543,14 @@ long IslandWindow::_calculateTotalSize(const bool isWidth, const long clientSize
...
+        auto lpwpos = (LPWINDOWPOS)lparam;
+
         // GH#10274 - if the quake window gets moved to another monitor via aero
         // snap (win+shift+arrows), then re-adjust the size for the new monitor.
         if (IsQuakeWindow())
         {
             // Retrieve the suggested dimensions and make a rect and size.
-            LPWINDOWPOS lpwpos = (LPWINDOWPOS)lparam;
-
             // We only need to apply restrictions if the position is changing.
```
Comment (miniksa) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:554]: This could be reverted.

Comment (miniksa) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:554]: This could be reverted.

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 5]
```diff
@@ -813,6 +819,15 @@ void IslandWindow::SetAlwaysOnTop(const bool alwaysOnTop)
     }
 }
 
+void IslandWindow::ShowWindowChanged(const bool showOrHide)
```
Comment (miniksa) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:822]: I don't have a doc comment unlike my friends.

Comment (miniksa) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:822]: I don't have a doc comment unlike my friends.

File: src/interactivity/base/InteractivityFactory.cpp [hunk 6]
```diff
@@ -5,6 +5,8 @@
 
 #include "InteractivityFactory.hpp"
 
+#include <dwmapi.h>
```
Comment (miniksa) [inline src/interactivity/base/InteractivityFactory.cpp:8]: Trim

Comment (miniksa) [inline src/interactivity/base/InteractivityFactory.cpp:8]: Trim

File: src/interactivity/base/InteractivityFactory.cpp [hunk 7]
```diff
@@ -306,18 +308,38 @@ using namespace Microsoft::Console::Interactivity;
...
+
                 if (hwnd == nullptr)
                 {
                     DWORD const gle = GetLastError();
                     status = NTSTATUS_FROM_WIN32(gle);
                 }
+
+                /*BOOL const cloak = TRUE;
+                DwmSetWindowAttribute(hwnd, DWMWA_CLOAK, &cloak, sizeof(cloak));*/
+
```
Comment (miniksa) [inline src/interactivity/base/InteractivityFactory.cpp:340]: Also trim. We won't use this here. It'll be added later to hide dialogs if we pan away from the tab.

Comment (miniksa) [inline src/interactivity/base/InteractivityFactory.cpp:340]: Also trim. We won't use this here. It'll be added later to hide dialogs if we pan away from the tab.

File: src/interactivity/base/InteractivityFactory.cpp (hunk 1 above)
Comment (miniksa) [reply src/interactivity/base/InteractivityFactory.cpp:418]: I don't care enough at this point. Leaving it here.

Comment (miniksa) [reply src/interactivity/base/InteractivityFactory.cpp:418]: I don't care enough at this point. Leaving it here.

**Review by zadjii-msft — COMMENTED** [review]
37/43

**Review by zadjii-msft — COMMENTED** [review]
okay I don't see any major concerns, other than the signal thing. That we should probably do.

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 8]
```diff
@@ -543,7 +548,7 @@ long IslandWindow::_calculateTotalSize(const bool isWidth, const long clientSize
         if (IsQuakeWindow())
         {
             // Retrieve the suggested dimensions and make a rect and size.
-            LPWINDOWPOS lpwpos = (LPWINDOWPOS)lparam;
+            LPWINDOWPOS lpwpos = (LPWINDOWPOS)lParam;
```
[the remark this answers is not in the document]
Comment (miniksa) [reply src/cascadia/WindowsTerminal/IslandWindow.cpp:551]: You're right oops.

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 9]
```diff
@@ -2789,6 +2802,35 @@ namespace winrt::TerminalApp::implementation
...
+    {
+        for (const auto& tab : _tabs)
+        {
+            if (auto terminalTab{ _GetTerminalTabImpl(tab) })
+            {
+                // Manually enumerate the panes in each tab; this will let us recycle TerminalSettings
+                // objects but only have to iterate one time.
+                terminalTab->GetRootPane()->WalkTree([&](auto&& pane) {
+                    if (const auto profile{ pane->GetProfile() })
+                    {
```
[the remark this answers is not in the document]
Comment (miniksa) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2823]: I thought you were the one who told me this is the way to do this?

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 10]
```diff
@@ -1683,6 +1692,22 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+    //         coming in via VT parsing (and rather from a window state transition)
+    //         we generate and send it here.
+    // Arguments:
+    // - visible: True for visible; false for not visible.
+    // Return Value:
+    // - <none>
+    void ControlCore::WindowVisibilityChanged(const bool showOrHide)
+    {
+        // show is true, hide is false
+        _connection.WriteInput(showOrHide ? L"\x1b[1t" : L"\x1b[2t");
```
[the remark this answers is not in the document]
Comment (miniksa) [reply src/cascadia/TerminalControl/ControlCore.cpp:1708]: Yeah so I initially thought I was going to have this go the other way with ConPTY querying the Terminal for its window state periodically and then I realized that was dumb and changed tacks to this strategy. But then I didn't switch over to a signal. You're right it probably should be a signal.

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2823]: I probably was 😝

**Comment by zadjii-msft** [comment]
I did the signal thing. This PR is a little awkward now, since I'll be maintaining it, but Michael authored it, so I don't get to count as a sign off _(right?)_

File: src/winconpty/winconpty.cpp [hunk 11]
```diff
@@ -459,6 +479,21 @@ extern "C" HRESULT WINAPI ConptyClearPseudoConsole(_In_ HPCON hPC)
...
+// - For more information, refer to GH#12515.
+extern "C" HRESULT WINAPI ConptyShowHidePseudoConsole(_In_ HPCON hPC, bool show)
+{
+    const PseudoConsole* const pPty = (PseudoConsole*)hPC;
+    HRESULT hr = pPty == nullptr ? E_INVALIDARG : S_OK;
+    if (SUCCEEDED(hr))
+    {
+        hr = _ShowHidePseudoConsole(pPty, show);
+    }
+    return hr;
```
Comment (lhecker) [inline src/winconpty/winconpty.cpp:494]: This entire thing can be replaced with:
```cpp
return _ShowHidePseudoConsole((PseudoConsole*)hPC, show);
```

Same for the function above that uses `_ClearPseudoConsole` btw.

**Review by lhecker — COMMENTED** [review]
The code looks alright. Some nits.

I didn't read the spec yet.

**Comment by zadjii-msft** [comment]
> I didn't read the spec yet.

TBH I think we're just gonna commit it at this point. I haven't touched it.

**Review by lhecker — APPROVED** [review]

**Comment by zadjii-msft** [comment]
@msftbot merge this in 4 hours

**Pull request merged**